src/storage/mod.rs

```diff
@@ -1073,6 +1194,83 @@ impl Storage {
             .flatten()
     }
 
+    pub fn async_raw_batch_scan(
+        &self,
+        ctx: Context,
+        mut ranges: Vec<KeyRange>,
+        each_limit: usize,
+        key_only: bool,
+    ) -> impl Future<Item = Vec<Result<KvPair>>, Error = Error> {
+        const CMD: &str = "raw_batch_scan";
+        let engine = self.get_engine();
+        let priority = readpool::Priority::from(ctx.get_priority());
+
+        let res = self.read_pool.future_execute(priority, move |ctxd| {
+            let mut _timer = {
+                let ctxd = ctxd.clone();
+                let mut thread_ctx = ctxd.current_thread_context_mut();
+                thread_ctx.start_command_duration_timer(CMD, priority, true)
+            };
+
+            Self::async_snapshot(engine, &ctx)
+                .and_then(move |snapshot: Box<Snapshot>| {
+                    let mut thread_ctx = ctxd.current_thread_context_mut();
+                    let _t_process = thread_ctx.start_processing_read_duration_timer(CMD);
+
+                    let mut statistics = Statistics::default();
+                    let mut result = Vec::new();
+                    let ranges_len = ranges.len();
+                    for i in 0..ranges_len {
+                        let start_key = Key::from_encoded(ranges[i].take_start_key());
+                        let end_key = ranges[i].take_end_key();
+                        let end_key = if end_key.is_empty() {
+                            if i + 1 == ranges_len {
+                                None
+                            } else {
+                                Some(Key::from_encoded(ranges[i + 1].get_start_key().to_vec()))
```

Comment:
do we need to check the next start key is bigger than the previous start key or end key?

Reply:
To make it more offensive, it should be. However I see other interfaces didn't even check the value exists. So do we agree at this moment client should be trustworthy or it's time to consider broken clients now. 

Anyway, let me add check first before start hardening others.

Reply:
Validation added, CI is going on.